src/config.rs

```diff
@@ -149,45 +157,135 @@ impl TlsClock for NoClock {
 }
 
 #[derive(Debug)]
-#[cfg_attr(feature = "defmt", derive(defmt::Format))]
-pub struct TlsContext<'a, CipherSuite, RNG>
+pub struct Signature<T: PrimeCurve>
 where
-    CipherSuite: TlsCipherSuite,
-    RNG: CryptoRng + RngCore + 'a,
+    SignatureSize<T>: core::ops::Add<ecdsa::der::MaxOverhead> + ArrayLength<u8>,
+    ecdsa::der::MaxSize<T>: ArrayLength<u8>,
 {
-    pub(crate) config: &'a TlsConfig<'a, CipherSuite>,
-    pub(crate) rng: &'a mut RNG,
+    signature: ecdsa::der::Signature<T>,
 }
 
-impl<'a, CipherSuite, RNG> TlsContext<'a, CipherSuite, RNG>
+impl<T: PrimeCurve> Deref for Signature<T> 
 where
-    CipherSuite: TlsCipherSuite,
-    RNG: CryptoRng + RngCore + 'a,
+    SignatureSize<T>: core::ops::Add<ecdsa::der::MaxOverhead> + ArrayLength<u8>,
+    ecdsa::der::MaxSize<T>: ArrayLength<u8>,
 {
-    /// Create a new context with a given config and random number generator reference.
-    pub fn new(config: &'a TlsConfig<'a, CipherSuite>, rng: &'a mut RNG) -> Self {
-        Self { config, rng }
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        self.signature.as_bytes()
     }
 }
 
-impl<'a, CipherSuite> TlsConfig<'a, CipherSuite>
+pub struct Signer<'a, T: PrimeCurve, RNG: CryptoRngCore> {
+    pub secret_key: SecretKey<T>,
+    pub scheme: SignatureScheme,
+    pub rng: &'a mut RNG,
+}
+
+impl<'a, T: PrimeCurve, RNG: CryptoRngCore> Signer<'a, T, RNG>
 where
-    CipherSuite: TlsCipherSuite,
+    SignatureSize<T>: core::ops::Add<ecdsa::der::MaxOverhead> + ArrayLength<u8>,
+    ecdsa::der::MaxSize<T>: ArrayLength<u8>,
 {
+    pub fn sign(&mut self, message: &[u8]) -> Signature<T>
+    where
+        T: CurveArithmetic + DigestPrimitive,
+        <T as CurveArithmetic>::Scalar: SignPrimitive<T>,
+    {
+        let signing_key = SigningKey::from(&self.secret_key);
+        let signature = signing_key.sign_with_rng(self.rng, &message);
+
+        Signature { signature }
+    }
+
+    pub fn signature_scheme(&self) -> SignatureScheme {
+        self.scheme
+    }
+}
+
+pub trait CryptoProvider {
+    type CipherSuite: TlsCipherSuite;
+    type SecureRandom: CryptoRngCore;
+    type SignatureCurve: CurveArithmetic + DigestPrimitive;
+
+    fn rng(&mut self) -> &mut Self::SecureRandom;
+
+    fn verifier(&mut self) -> Result<&mut impl TlsVerifier<Self::CipherSuite>, crate::TlsError> {
+        Err::<&mut NoVerify, _>(crate::TlsError::Unimplemented)
+    }
+
+    /// Decode and validate a private signing key from `key_der`.
+    fn signer(
+        &mut self,
+        _key_der: &[u8],
+    ) -> Result<Signer<Self::SignatureCurve, Self::SecureRandom>, crate::TlsError> {
+        Err::<Signer<Self::SignatureCurve, Self::SecureRandom>, _>(crate::TlsError::Unimplemented)
+    }
+}
+
+pub struct SimpleProvider<CipherSuite, RNG> {
```

Comment:
Don't know if anyone has a better name for this that `SimpleProvider`? It really is just a means to instantiate a `CryptoProvider` in all the cases where no `Verifier` nor `Signer` is needed, by just passing an RNG source and selecting a Cipher suite

Reply:
UnsecureProvider maybe? I think it's somewhat important to convey that it doesn't do any checking at least.